optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java

```diff
@@ -0,0 +1,219 @@
+                .filter(((machineCapacity, usage) -> machineCapacity.getSafetyCapacity() < usage))
+                .penalizeLong((machineCapacity, usage) -> machineCapacity.getSafetyCapacity() - usage);
+    }
+
+    /**
+     * availability(r) = capacity(m, r) - usage(m, r)
+     * balanceCost = sum(max(0, multiplier * availability(m, r1) - availability(m, r2)))
+     */
+    private void balanceCost(ConstraintFactory constraintFactory) {
+        throw new UnsupportedOperationException("Not yet implemented due to missing support for quad streams.");
+
+        /*
+        Constraint constraint = constraintFactory.newConstraintWithWeight(MrConstraintName.BALANCE_COST.getName(),
+                                                                          HardSoftLongScore.ofSoft(1L));
+        constraint.from(MrBalancePenalty.class)
+                .join(MrProcessAssignment.class)
+                .groupBy((penalty) -> penalty, (penalty, assignment) -> assignment.getMachine(),
+                         sumLong((penalty, assignment) -> processAssignment.getUsage(penalty.getOrigin())
+                                 sumLong((penalty, assignment) ->
+                                                 processAssignment.getUsage(penalty.getTarget()))))
+                // QuadStream<MrBalancePenalty, MrMachine, Long, Long>
+                .filter((penalty, machine, originUsage, targetUsage) ->
+                                (machine.getCapacity(penalty.getOrigin()) - originUsage) <
+                                        penalty.getMultiplier() * (machine.getCapacity(penalty.getTarget()) - targetUsage))
+
+         */
+    }
+
+    /**
+     * Process move cost: A process has a move cost.
+     */
+    private void processMoveCost(ConstraintFactory constraintFactory) {
+        Constraint constraint = constraintFactory.newConstraintWithWeight(MrConstraintName.PROCESS_MOVE_COST.getName(),
+                                                                          HardSoftLongScore.ofSoft(1L));
+        constraint.from(MrProcessAssignment.class)
+                .filter(processAssignment -> processAssignment.isMoved())
+                .penalizeLong(MrProcessAssignment::getProcessMoveCost);
+    }
+
+    /**
+     * Service move cost: A service has a move cost.
+     */
+    private void serviceMoveCost(ConstraintFactory constraintFactory) {
+        throw new UnsupportedOperationException("Not yet implemented due to missing aggregation function.");
+
+        /*
+        Constraint constraint = constraintFactory.newConstraintWithWeight(MrConstraintName.SERVICE_MOVE_COST.getName(),
+                                                                          HardSoftLongScore.ofSoft(1L));
+        constraint.from(MrProcessAssignment.class)
+                .filter(processAssignment -> processAssignment.isMoved())
+                .groupBy(processAssignment -> processAssignment.getService(), ConstraintCollectors.count())
+                .penalizeLong();
```

Comment:
This is missing a max or something, right? Let's add a TODO there so it sticks out